### ISSUE TITLE
pbalign call blasr with a non supported --sam option

### ISSUE THREAD

**Issue by RxLoutre** [PacificBiosciences/pbalign#48]
Hi everyone !

I'm trying to use pbalign as a first step to polish my assembly with Quiver. I've made a FOFN file of all my PacBio raw reads, and then I launched pbalign this way : 
`
pbalign --forQuiver /media/loutre/SUZUKII/assembly_tries/polishing/raw_reads.fofn /media/loutre/SUZUKII/assembly_tries/cleaned_canu3_suzukii.fasta /media/loutre/SUZUKII/assembly_tries/polishing/canu3_cleaned_aligned_suzukii.cmp.h5`

As pbalign call blasr, I think that somehow the blasr version I have and the GenomicConsensus version I have (the first one, not the second one ) are not compatible. Here the message error I get : 

```
[INFO] 2016-08-12 13:41:35,094Z [root run 255] pbalign version: 0.2.0.173392
[INFO] 2016-08-12 13:41:35,737Z [root run 178] BlasrService: Align reads to references using blasr.
[INFO] 2016-08-12 13:41:35,737Z [root Execute 63] BlasrService: Call "blasr /media/loutre/SUZUKII/assembly_tries/polishing/raw_reads.fofn /media/loutre/SUZUKII/assembly_tries/cleaned_canu3_suzukii.fasta --out /tmp/tmpm2ecFT/K_y5Qr.sam  --sam  --bestn 10 --minMatch 12  --nproc 8  --minSubreadLength 50 --minAlnLength 50  --minPctSimilarity 70 --minPctAccuracy 70 --hitPolicy randombest  --randomSeed 1 "
BlasrService returned a non-zero exit status. [INFO] 2016-08-12T15:41:35 [blasr] started.
ERROR: --sam is no longer supported, use --bam, then translate from .bam to .sam
[ERROR] 2016-08-12 13:41:35,743Z [root Execute 67] BlasrService returned a non-zero exit status. [INFO] 2016-08-12T15:41:35 [blasr] started.
ERROR: --sam is no longer supported, use --bam, then translate from .bam to .sam
[INFO] 2016-08-12 13:41:35,743Z [pbalign.pbalignrunner _pacbio_main_runner 153] exiting with return code 1 in 0.65 sec.
```

Should I just modify inside the pbalign caller that launch blasr and convert "--sam" in a "--bam" ?

Thanks for your advices !

Cheers,

Roxane

**Comment by vrainish-pacbio** [comment]
Roxane, 

We recently deprecated --sam option. I need to make this small fix in pbalign. Hope to make it today.

Vladimir

**Comment by RxLoutre** [comment]
Thanks Vladimir, can you please notice me when this problem is solved ?
Also help me a little bit with what should I do to take account of the change you will make, you know that sometimes I could be a bit slow, but "vive la persévérance" !

**Comment by RxLoutre** [comment]
I'm not sure, but I think that I can just modify theses lines in pbalign/alignservice/blasr.py : 
```

 if getFileFormat(fileNames.alignerSamOut) == FILE_FORMATS.BAM:
            cmdStr += " --bam "
        else:
            cmdStr += " --sam "
```

**Comment by vrainish-pacbio** [comment]
Hi Roxane,

My apologies for not answering, I have been swamped by other tasks. This issue on definitely my list of fixes, it might be a little bit more ( not really much more) then switch you added. I'll try to look into it tomorrow and notify you how it's going.

Regards, Vladimir

**Comment by RxLoutre** [comment]
Hi Vladimir,

Sorry for my impatience, but I really need to use pbalign, I hope that the issue will be solved soon and that I can try Quiver in the next week !

Cheers,

Roxane

**Comment by pb-cdunn** [comment]
Vlad, is this in our current sprint?

**Comment by vrainish-pacbio** [comment]
It’s on my todo list anyway. I’ll do it today later

**Comment by pb-cdunn** [comment]
Vlad! At least create a task in the backlog, even if you finish it before it goes into a sprint. Let this kind of thing be tracked somehow, ok?

**Comment by vrainish-pacbio** [comment]
No problems

**Comment by vrainish-pacbio** [comment]
Hi Chris,

I’ll create it, there is an existing task in Bugzilla I’ll migrate.

Regards

**Comment by RxLoutre** [comment]
Hi Vladimir, Christopher,

Could you please notify me as soon as possible when you will fix it ?

Thanks,

Roxane

**Comment by vrainish-pacbio** [comment]
Hi Roxane,

Sorry fir delay with this, I was down with cold most of the weekend and Monday. Will try to fix it tomorrow.

**Comment by RxLoutre** [comment]
Hi Vladimir,

Hope you will feel better soon !
Take care,

Roxane

**Comment by RxLoutre** [comment]
Dear Vladimir,

I've been waiting for more than two weeks now.

I hope you are doing okay and that I could try pbalign and Quiver soon.

Cheers,

Roxane

**Comment by vrainish-pacbio** [comment]
Hi Roxane ,

I needed to complete an urgent assignment over the last week and today ( skipping some time because of cold last week didn't help). I am nearly done and I'll try to fix it tomorrow ( can't promise, but try my very best).

To make it easier, please, confirm the following . It's very important since it can save me time

0.  You've cloned reasonably current **pitchfork** ( not older then couple of weeks ago )

1. You've succesfully build blasr with pbbam support and pbalign **using pitchfork**

Denote that pitchfork builds blasr with pbbam support by default, 
so you don't need to specify any additional flags simply 
**make blasr** 
and 
**make pbalign**
should be enough

2. You can run **blasr --version** 

It's very important , since in the past you experienced issues there because of LD_LIBRARY_PATH but those were overcome

Let me know, and as I mentioned I'll try to look into it tomorrow.

**Comment by RxLoutre** [comment]
Hi Vladimir,

As pitchwork was a nightmare to install (even with the help of the IT people in my lab), I managed to install myself all PacBio tools manually.

I have blasr, GenomicConsensus, ConsensusCore... Everything that is needed by Quiver.

I can indeed run blasr : 

```
Loutre:~/Documents/maker2$ blasr --version
blasr	5.2.def62de
Loutre:~/Documents/maker2$ echo $LD_LIBRARY_PATH
/home/loutre/blasr/libcpp/alignment:/home/loutre/blasr/libcpp/hdf:/home/loutre/blasr/libcpp/pbdata:/home/loutre/hdf5/lib::::
```

I have solved the problem about this. The problem I have is when running pb align. it seems that pbalign call blasr with a non supported option --bam (or sam can't remember)

**Comment by pawelosipowski** [comment]
Hi, I installed pbalign and blasr through pitchfork, ran those in env as instructed but get the same problem:
`BlasrService returned a non-zero exit status. [INFO] 2016-09-14T08:50:22 [blasr] started.
ERROR: --sam is no longer supported, use --bam, then translate from .bam to .sam
[ERROR] 2016-09-14 06:50:22,208Z [root Execute 67] BlasrService returned a non-zero exit status. [INFO] 2016-09-14T08:50:22 [blasr] started.
ERROR: --sam is no longer supported, use --bam, then translate from .bam to .sam
[INFO] 2016-09-14 06:50:22,209Z [pbalign.pbalignrunner _pacbio_main_runner 153] exiting with return code 1 in 0.33 sec.`

Can we count on some updates?

**Comment by vrainish-pacbio** [comment]
Run
blasr  --version
Please post the output

**Comment by vrainish-pacbio** [comment]
Hi Roxane,

Sorry for late reply, Last few days where dedicated to get --sam option back to Blasr ( see my previous post for some details) and it's now in blasr 5.3 version which I submitted yesterday . I asked you to ensure you can build blasr with pitchfork for a reason : it must be built with pbbam , building it with pbbam in stand-alone mode is much more complicated, since you need to install a certain version of Cmake and also clone/built pbbam itself, Boost libraries and pacbvio version of hstlib.

You wrote *As pitchwork was a nightmare to install (even with the help of the IT people in my lab),*.
I can't understand why ( unless you have a customized OS but then everything might be nightmare).
With pitchfork you just clone it , modify couple of liness in **settngs.mk**  and then run make blasr, make pbalign etc. etc. many people outside of Pacbio used it succesfully. (comparing with it, installing blasr with pbbam, where you need to do everything manually, would be a real nightmare).

Anyway Blasr now *should* accept --sam. You would like to give it a try, let me know and I'll try to help. You have problem with pitchfork install, open an issue on pitchfork, as I mentioned, many people have been using it successfully, and if they had initial problems, they were resolved.   

Please, realize that if you have a non-standard install of OS and/or version of gcc we don't support, you might have potential problems ( we can't ensure our environment runs on EVERY variant of Linux and on every version of gcc) 

Regards, Vladimir

**Comment by pawelosipowski** [comment]
I have to ask it here as I'm desperate. It seems that after launching your setup-env.sh with bash I terminated terminal window and it modified system on start up. Details are here:http://unix.stackexchange.com/questions/309891/no-access-to-system-binaries-after-switching-off-terminal-with-bash-session-with. Can you help me?

**Comment by vrainish-pacbio** [comment]
Hi Pawel,

It looks like the problem you've just reported has nothing to do with --sam option of blasr, but  related to general install/configuration. I suggest you open an issue on pitchfork repository.

Once it's solved, please, ensure you have the latest version of blasr (run **blasr --version**  which should print **5.3....**) then you can attempt to rerun your tasks with **--sam** option 

Vladimir

**Comment by pawelosipowski** [comment]
Hi Vladimir,
 thanks, I'll put it there. Thanks for --sam work!
Pawel

-----------------------------------------------------------------------------------------------------------

Paweł Osipowski - Ph.D. candidate, Warsaw University of Life Sciences,
Department of Plant Genetics Breeding and Biotechnology, 159
Nowoursynowska Street 02-776 Warsaw, Poland, phone: +48 22 5932167,
fax: +48 22 5932152, e-mail: pawel
<grzegorz_bartoszewski@sggw.pl>.sggw@gmail.com


2016-09-15 21:46 GMT+02:00 Vladimir Rainish <notifications@github.com>:

**Comment by mhsieh** [comment]
@RxLoutre hi I would love to help out, if you find pitchfork hard to work with, can you guys file an issue in pitchfork project?

**Comment by RxLoutre** [comment]
Hi Vladimir,

Sorry for the long answers, I was back to school, and I couldn't access to my work computer. 

Now that I'm back to work, I tried to re-install blasr and hdf5, and then I relaunched my pbalign command.

The message have a bit changed, but it still give me the same error.

Here is my command : 

`pbalign --forQuiver /media/loutre/SUZUKII/assembly_tries/polishing/raw_reads.fofn /media/loutre/SUZUKII/assembly_tries/pb_268_p_ctg.fasta /media/loutre/SUZUKII/assembly_tries/polishing/falcon_suzukii.cmp.h5`

Here is a part of the results it gives as output (I don't show everything because it print twice the help of blasr now ) : 

`[INFO] 2017-01-09 09:23:11,232Z [pbalign.pbalignrunner _pacbio_main_runner 128] Using pbcommand v0.3.29
[INFO] 2017-01-09 09:23:11,232Z [pbalign.pbalignrunner _pacbio_main_runner 129] completed setting up logger with <function setup_log at 0x7f98e28d2de8>
[INFO] 2017-01-09 09:23:11,233Z [pbalign.pbalignrunner _pacbio_main_runner 130] log opts {'file_name': None, 'level': 20}
[INFO] 2017-01-09 09:23:11,233Z [root run 255] pbalign version: 0.2.0.173392
[INFO] 2017-01-09 09:23:12,166Z [root run 178] BlasrService: Align reads to references using blasr.
[INFO] 2017-01-09 09:23:12,166Z [root Execute 63] BlasrService: Call "blasr /media/loutre/SUZUKII/assembly_tries/polishing/raw_reads.fofn /media/loutre/SUZUKII/assembly_tries/pb_268_p_ctg.fasta --out /tmp/tmpR1V25p/WFJYKu.sam  --sam  --bestn 10 --minMatch 12  --nproc 8  --minSubreadLength 50 --minAlnLength 50  --minPctSimilarity 70 --minPctAccuracy 70 --hitPolicy randombest  --randomSeed 1 "
BlasrService returned a non-zero exit status.    Options for blasr 
   Basic usage: 'blasr reads.{bam|fasta|bax.h5|fofn} genome.fasta [-options] 
 option	Description (default_value).
.... Help of blasr ....
In release v5.1 of BLASR, command-line options will use the 
single dash/double dash convention: 
Character options are preceded by a single dash. (Example: -v) 
Word options are preceded by a double dash. (Example: --verbose) 
Please modify your scripts accordingly when BLASR v5.1 is released. 

To cite BLASR, please use: Chaisson M.J., and Tesler G., Mapping 
single molecule sequencing reads using Basic Local Alignment with 
Successive Refinement (BLASR): Theory and Application, BMC 
Bioinformatics 2012, 13:238.
Please report any bugs to 'https://github.com/PacificBiosciences/blasr/issues'.


ERROR: --sam is not a valid option.
[ERROR] 2017-01-09 09:23:12,173Z [root Execute 67] BlasrService returned a non-zero exit status.    Options for blasr 
   Basic usage: 'blasr reads.{bam|fasta|bax.h5|fofn} genome.fasta [-options] 
 option	Description (default_value).
....Second time help of blasr......
In release v5.1 of BLASR, command-line options will use the 
single dash/double dash convention: 
Character options are preceded by a single dash. (Example: -v) 
Word options are preceded by a double dash. (Example: --verbose) 
Please modify your scripts accordingly when BLASR v5.1 is released. 

To cite BLASR, please use: Chaisson M.J., and Tesler G., Mapping 
single molecule sequencing reads using Basic Local Alignment with 
Successive Refinement (BLASR): Theory and Application, BMC 
Bioinformatics 2012, 13:238.
Please report any bugs to 'https://github.com/PacificBiosciences/blasr/issues'.


ERROR: --sam is not a valid option.
[INFO] 2017-01-09 09:23:12,177Z [pbalign.pbalignrunner _pacbio_main_runner 153] exiting with return code 1 in 0.94 sec.
`
Of course, I have checked the blasr version : 

`Loutre:~$ blasr --version
blasr	5.3.8d086d7`

So, as I have the latest version of blasr, but pbalign keep talking about version 5.1, what's the problem here ? Do you have any idea ?

Thanks for your help,

Roxane

PS : I do not use pitchfork, I've manually installed everything

**Comment by RxLoutre** [comment]
Hi again,

As it seems very important to use within pitchfork, I tried again to install all the tools using pitchfork, but here is problem.

Inside pitchfork, when I use make on some target, it work (for example, blasr was installed correctly), but when i tried to install pbalign, I got this error at this end of the installation : 

`Generating openblas_config.h in /home/loutre/pitchfork/staging/OpenBLAS-0.2.18/include
Generating f77blas.h in /home/loutre/pitchfork/staging/OpenBLAS-0.2.18/include
Generating cblas.h in /home/loutre/pitchfork/staging/OpenBLAS-0.2.18/include
Copying LAPACKE header files to /home/loutre/pitchfork/staging/OpenBLAS-0.2.18/include
Copying the static library to /home/loutre/pitchfork/staging/OpenBLAS-0.2.18/lib
Copying the shared library to /home/loutre/pitchfork/staging/OpenBLAS-0.2.18/lib
install: impossible d'évaluer «libopenblas_sandybridgep-r0.2.18.so»: Aucun fichier ou dossier de ce type`

It seems that, each time, it couldn't read some required libraires. I didn't used any PREFIX to build my targets.

Did I made something wrong ?

Cheers,

Roxane
